lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -171,9 +172,273 @@ bool swift::isIntermediateRelease(SILInstruction *inst,
+/// operands of the instruction and requires cleaning up the end-of-scope uses
+/// of non-trivial results.
+///
+/// \param inst instruction that checked for liveness.
+/// \param knownPureValueTypes a set of known non-trivial types that are known
+/// to have pure value value semantics. These types are populated when a
+/// constant_evaluable call is seen, as it would accept only types with value
+/// semantics.
+static bool
+isScopeAffectingInstructionDead(SILInstruction *inst,
+                                SmallPtrSetImpl<CanType> &knownPureValueTypes) {
+  SILFunction *fun = inst->getFunction();
+  assert(fun && "Instruction has no function.");
+  // Only support ownership SIL for scoped instructions.
+  if (!fun->hasOwnership()) {
+    return false;
+  }
+  // If the instruction has any use other than end of scope use or destroy_value
+  // use, bail out.
+  if (!hasOnlyEndOfScopeOrDestroyUses(inst)) {
+    return false;
+  }
+  // If inst is a copy or beginning of scope, inst is dead, since we know that
+  // it is used only in a destroy_value or end-of-scope instruction.
+  if (getSingleValueCopyOrCast(inst))
+    return true;
+
+  switch (inst->getKind()) {
+  case SILInstructionKind::LoadInst: {
+    LoadOwnershipQualifier loadOwnershipQual =
+        cast<LoadInst>(inst)->getOwnershipQualifier();
+    // If the load creates a copy, it is dead, since we know that if at all it
+    // is used, it is only in a destroy_value instruction.0
+    return (loadOwnershipQual == LoadOwnershipQualifier::Copy ||
+            loadOwnershipQual == LoadOwnershipQualifier::Trivial);
+  }
+  case SILInstructionKind::PartialApplyInst: {
+    // Partial applies that are only used in destroys cannot have any effect on
+    // the program state, provided the values they capture are explicitly
+    // destroyed.
+    return true;
+  }
+  case SILInstructionKind::StructInst:
+  case SILInstructionKind::EnumInst:
+  case SILInstructionKind::TupleInst:
+  case SILInstructionKind::ConvertFunctionInst:
+  case SILInstructionKind::DestructureStructInst:
+  case SILInstructionKind::DestructureTupleInst: {
+    // All these ownership forwarding instructions that are only used in
+    // destroys are dead provided the values they consume are destroyed
+    // explicitly.
+    return true;
+  }
+  case SILInstructionKind::ApplyInst: {
+    // A constant_evaluable call can only return a type with value semantics.
+    // The deinit of the returned type should be pure in the sense that it only
+    // affects its self and nothing else. Therefore, it should be safe to remove
+    // the call along with any destroys of the returned value, if the call
+    // itself doesn't have any side-effects i.e., if it is a read-only call.
+    // Nonetheless, the call is allowed to consume (i.e., destroy) the owned
+    // values passed to the call.
+    FullApplySite applySite(cast<ApplyInst>(inst));
+    if (isReadOnlyConstantEvaluableCall(applySite)) {
```

Comment:
Two interesting cases to spell out:
- the call destroys its arguments: when removing the call, a destroy of each argument will be added
- the call returns a dead value: both the call and corresponding destroy will be removed